### PR TITLE
Added POST /user/reset_password route

### PR DESCRIPTION
**Pull request by dholms**
## Problem
No way for users to reset their password

## Impact 
If a user exposes their password/api keys, then they can't cycle their credentials

## Solution
Add a POST /user/reset_password route
- optional password param
- if not included then it defaults to a random password

Closes https://github.com/fission-suite/web-api/issues/191

### CODE REVIEW

File: library/Fission/Web/User/Password/Reset.hs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import           Servant
+
+import           Fission.Web.Server
+import qualified Fission.Web.Error    as Web.Err
+
+import           Fission.User                 as User
+import qualified Fission.User.Password.Types  as User
+
+type API = ReqBody '[JSON] User.Password
+        :> Post '[JSON] User.Password
```
Comment (expede) [inline library/Fission/Web/User/Password/Reset.hs]: Per REST, this should be a `PUT`. I wish that they had named it better; `PUT` vs `POST` is confusing.

File: library/Fission/User/Mutate.hs [hunk 2]
```diff
@@ -71,6 +74,29 @@ create' username password email herokuUUID = do
       logInfo $ "Inserted user " <> display uID
       return $ Right uID
 
+updatePassword :: MonadRIO    cfg m
+               => MonadSelda      m
+               => HasLogFunc  cfg
+               => ID User
+               -> User.Password
+               -> m (Either Error.Create User.Password)
```
Comment (expede) [inline library/Fission/User/Mutate.hs]: 💅 Question for ya: how do you feel about this style:

```haskell
updatePassword 
  :: ( MonadRIO    cfg m
     , MonadSelda      m
     , HasLogFunc  cfg
     )
  => ID User
  -> User.Password
  -> m (Either Error.Create User.Password)
```

This is both a more common style, and after a few months and two onboardings, I think possibly clearer than the row of `=>`. Very clear where the constraints are versus arguments.

Dropping the `::` to the next line is newer, but starting to catch on in the community, too. It makes indentation more consistent.

I was thinking about us slowly changing to this style as we go. Any thoughts?

File: library/Fission/User/Mutate.hs [hunk 3]
```diff
@@ -71,6 +74,29 @@ create' username password email herokuUUID = do
...
+    Right secretDigest -> do
+      update Table.users 
+        (\user -> user ! #_userID .== literal userID)
+        (\user -> user `with` [#_secretDigest := literal secretDigest])
+      logInfo $ "Updated password for user " <> display userID
+      return . Right . User.Password $ Just password
+
+defaultPassword :: MonadRIO cfg m
+                => Maybe Text
+                -> m (Text)
```
Comment (expede) [inline library/Fission/User/Mutate.hs]: 💅

* No need for the brackets around `(Text)`
* No need for the brackets around `(Just pw)` in this case
* This doesn't need to be full `MonadRIO`; `MonadIO` will do
  * `MonadRIO` is for when you need to pass around some config

```haskell
defaultPassword :: MonadIO m => Maybe Text -> m Text
defaultPassword Just pw = return pw
defaultPassword Nothing = liftIO $ Random.text 200
```

File: library/Fission/Web/User/Password/Reset.hs (hunk 1 above)
Comment (dholms) [reply library/Fission/Web/User/Password/Reset.hs]: Yup you're right. I often mix those up

File: library/Fission/User/Mutate.hs (hunk 2 above)
Comment (dholms) [reply library/Fission/User/Mutate.hs]: Already left my thoughts in the other PR :wink: but yes big fan

File: library/Fission/User/Mutate.hs (hunk 3 above)
Comment (dholms) [reply library/Fission/User/Mutate.hs]: Good point on 1 & 3. Compiler freaks out without brackets around `Just pw` though: `Equations for ‘defaultPassword’ have different numbers of arguments`

**Pull request merged**